src/encoded/batch_download.py

```diff
@@ -229,6 +231,35 @@ def metadata_tsv(context, request):
         param_list['field'] = param_list['field'] + _tsv_mapping[prop]
         if _tsv_mapping[prop][0].startswith('files'):
             file_attributes = file_attributes + [_tsv_mapping[prop][0]]
+
+    # Handle metadata.tsv lines from cart-generated files.txt.
+    if 'cart' in param_list:
+        # metadata.tsv line includes cart UUID, so load the specified cart and
+        # get its "elements" property for a list of items to retrieve.
+        if len(param_list['cart']) != 1:
+            msg = 'Cannot have more than one "cart" query string parameter.'
+            raise HTTPBadRequest(explanation=msg)
+        cart_uuid = param_list['cart'][0]
+        del param_list['cart']
+        try:
+            cart = request.embed(cart_uuid, '@@object')
+        except KeyError:
+            msg = 'No cart with uuid {0}.'.format(cart_uuid)
+            raise HTTPBadRequest(explanation=msg)
+        if len(cart['elements']) > 0:
+            param_list['@id'] = cart['elements']
+    else:
+        # If the metadata.tsv line includes a JSON payload, get its "elements"
+        # property for a list of items to retrieve.
+        try:
+            elements = request.json.get('elements')
+        except ValueError:
```

Comment:
`request.json.get('elements')` raises a ValueError if no JSON object exists in the request (such as for a regular files.txt file from the search page), causing this function to generate a metadata.tsv with an HTML error message as its content. If a JSON object exists but has no "elements" property, then you’re right — we get `param_list['@id']=None` which then gets passed on to the query string, so I fixed that.

Looking at this section revealed something else: it’s not a good idea to raise anything in the middle of `metadata_tsv` because it results in the generation of an _HTML_ file called “metadata.tsv” which gets really awkward. So I changed the code to never raise, but it results in some extra ifs that I’m not sure you’ll like.